Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp

```diff
@@ -147,10 +209,108 @@ RefPtr<TypeDefinition> TypeDefinition::tryCreateStructType(StructFieldCount fiel
+    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(TypeDefinitionKind::StructType, fieldCount);
+    return adoptRef(signature);
+}
+
+RefPtr<TypeDefinition> TypeDefinition::tryCreateRecursionGroup(RecursionGroupCount typeCount)
+{
+    // We use WTF_MAKE_FAST_ALLOCATED for this class.
+    auto result = tryFastMalloc(allocatedRecursionGroupSize(typeCount));
+    void* memory = nullptr;
+    if (!result.getValue(memory))
+        return nullptr;
+    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(TypeDefinitionKind::RecursionGroup, typeCount);
+    return adoptRef(signature);
+}
+
+RefPtr<TypeDefinition> TypeDefinition::tryCreateProjection()
+{
+    // We use WTF_MAKE_FAST_ALLOCATED for this class.
+    auto result = tryFastMalloc(allocatedProjectionSize());
+    void* memory = nullptr;
+    if (!result.getValue(memory))
+        return nullptr;
+    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(TypeDefinitionKind::Projection);
     return adoptRef(signature);
 }
 
+// Recursive types are stored "tied" in the sense that the spec refers to here:
+//
+//   https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#equivalence
+//
+// That is, the recursive "back edges" are stored as a special type index. These
+// need to be substituted back out to a Projection eventually so that the type
+// can be further expanded if necessary. The substitute and replacePlaceholders
+// functions below are used to implement this substitution.
+Type TypeDefinition::substitute(Type type, TypeIndex projectee)
+{
+    if (type.kind == TypeKind::Rec) {
+        RefPtr<TypeDefinition> projection = TypeInformation::typeDefinitionForProjection(projectee, static_cast<ProjectionIndex>(type.index));
+        TypeKind kind = type.isNullable() ? TypeKind::RefNull : TypeKind::Ref;
+        return Type { kind, type.nullable, projection->index() };
+    }
+
+    return type;
+}
+
+// This operation is a helper for expand() that calls substitute() in order
+// to replace placeholder recursive references in structural types.
+const TypeDefinition& TypeDefinition::replacePlaceholders(TypeIndex projectee) const
+{
+    if (is<FunctionSignature>()) {
+        const FunctionSignature* func = as<FunctionSignature>();
+        Vector<Type> newArguments;
+        Vector<Type, 1> newReturns;
+        newArguments.tryReserveCapacity(func->argumentCount());
+        newReturns.tryReserveCapacity(func->returnCount());
+        for (unsigned i = 0; i < func->argumentCount(); ++i)
+            newArguments.uncheckedAppend(substitute(func->argumentType(i), projectee));
+        for (unsigned i = 0; i < func->returnCount(); ++i)
+            newReturns.uncheckedAppend(substitute(func->returnType(i), projectee));
+
+        RefPtr<TypeDefinition> def = TypeInformation::typeDefinitionForFunction(newReturns, newArguments);
+        return *def;
+    }
+
+    if (is<StructType>()) {
```

Comment:
Added this `StructType` case that was previously unimplemented (I thought it wasn't exercised yet, but existing tests already get to this case it turns out).